typescript/test/bitcoin.test.ts

```diff
@@ -45,4 +50,145 @@ describe("Bitcoin", () => {
+              const wrontPublicKeyHash = "02" + publicKeyHash
+
+              expect(() =>
+                encodeToBitcoinAddress(wrontPublicKeyHash, true, "main")
+              ).to.throw()
+            })
+          })
+        })
+
+        context("when witness option is false", () => {
+          context("when proper public key hash is provided", () => {
+            it("should encode public key hash into bitcoin address properly", () => {
+              expect(
+                encodeToBitcoinAddress(publicKeyHash, false, "main")
+              ).to.be.equal(P2PKHAddress)
+            })
+          })
+
+          context("when wrong public key hash is provided", () => {
+            it("should throw", () => {
+              const wrongPublicKeyHash = "02" + publicKeyHash
+
+              expect(() =>
+                encodeToBitcoinAddress(wrongPublicKeyHash, false, "main")
+              ).to.throw()
+            })
+          })
+        })
+      })
+    })
+
+    describe("decodeAddress", () => {
+      context("when network is main", () => {
+        context("when proper P2WPKH address is provided", () => {
+          it("should decode P2WPKH adress correctly", () => {
+            expect(decodeBitcoinAddress(P2WPKHAddress)).to.be.equal(
+              publicKeyHashTestnet
+            )
+          })
+        })
+
+        context("when proper P2PKH address is provided", () => {
+          it("should decode P2PKH address correctly", () => {
+            expect(decodeBitcoinAddress(P2PKHAddress)).to.be.equal(
+              publicKeyHashTestnet
+            )
+          })
+        })
+
+        context("when wrong address is provided", () => {
+          it("should throw", () => {
+            const bitcoinAddress = "123" + P2PKHAddress
+
+            expect(() => decodeBitcoinAddress(bitcoinAddress)).to.throw()
+          })
+        })
+      })
+
+      context("when network is testnet", () => {
+        context("when proper P2WPKH address is provided", () => {
+          it("should decode P2WPKH adress correctly", () => {
+            expect(decodeBitcoinAddress(P2WPKHAddressTestnet)).to.be.equal(
+              publicKeyHashTestnet
+            )
+          })
+        })
+
+        context("when proper P2PKH address is provided", () => {
+          it("should decode P2PKH address correctly", () => {
+            expect(decodeBitcoinAddress(P2PKHAddressTestnet)).to.be.equal(
+              publicKeyHashTestnet
+            )
+          })
+        })
+
+        context("when wrong address is provided", () => {
+          it("should throw", () => {
+            const bitcoinAddress = "123" + P2PKHAddressTestnet
+
+            expect(() => decodeBitcoinAddress(bitcoinAddress)).to.throw()
+          })
+        })
+      })
+    })
+  })
+
+  describe("isPubkeyhash", () => {
+    const publicKeyHash = "3a38d44d6a0c8d0bb84e0232cc632b7e48c72e0e"
+    const witnessPublicKeyHash = "3a38d44d6a0c8d0bb84e0232cc632b7e48c72e0e"
```

Comment:
Additional to above comments here is a good example of what i mean. We use exactly the same public key hash to check if `isPubkeyhash` works properly with `witness` option set to `true` and to `false` and in both scenarios it returns a `true` value.